### DISCUSSION TITLE
[Show & Tell] Finding the right component using smart LEDs

### DISCUSSION THREAD

**Discussion by Zontex** [inventree/InvenTree#2678] [Show and tell]
Hi Inventree team,

I Just would like to share with you what I've been working on recently, I've hosted the Inventree on top of Raspberry Pi CM4 and have been using it for the last couple of months as my main inventory management system (all my components and everything is managed through Inventree! couldn't live without it).

It's deployed on bare metal custom hardware I designed called "OmniKit" which has lots of features including 4G/LTE module, I've made a demonstration where when I take many components for testing and I want to return them back, instead of looking where to return I just scan the barcode / QR code and it returns feedback by LEDs, check this out: 
https://www.reddit.com/r/raspberry_pi/comments/t0c99r/raspberry_pi_cm4_smart_inventory_management/

Another thing id like to make next is SMS notifications when a product is taken / returned to the inventory, I think this could be very useful. right now it's bare python scripts but I should probably make plugins out of it to contribute back to the community, let me know what do you think and if any of this features could be further developed by me to enhance the Inventree community :)

**Comment by SchrodingersGat** [comment]
@Zontex that's amazing! We've been discussing something like this [for ages](https://github.com/inventree/InvenTree/issues/749). I'm very pleased that you are making such good use of the InvenTree software!

This would make a great plugin (using the new plugin framework by @matmair)! Would you be willing to support something like this?

Regarding SMS notifications - the notification framework is undergoing an overhaul / refactor for the next stable release. @matmair is leading the charge on this one. I'm not sure whether SMS works best as a "core feature" or a plugin

**Comment by Zontex** [comment]
> 

Regarding the LED I would definitely support it, I figured out I should design a custom strip wireless supporter controlled by either HTTP and/or MQTT instead of connecting the strip directly to Raspberry Pi (OmniKit in my case) I already designed the logic where each LED is appointed to specific shelf ID so it can be lighten up properly, I think this should be used as external plugin, I would love to continue contributing to this feature by hardware and firmware and perhaps some python scripts like I do now.

Regarding the SMS, it would be nice to embedded it as a core feature with option to set a hook API for twillio service or custom service (can make a flask service to integrate to OmniKit 4G EC25 module to send SMS by rest API) by creating it as a core with twillio/custom api hook it will be very easy to integrate and create different types of messages for different scenerios.

perhaps, the SMS system could be general notification as if it goes to external API this could be used for integration with whatsapp / telegram bot which will be more cost effective soltuion.

**Comment by matmair** [comment]
@Zontex rest/mqtt integration is very easily possible. Interacting with GPIO would also be easy. If you have questions ping me.

The new notification method is generic. So any plugin can trigger a notification and the delivery methods are defined per user. SMS, Telegram or pidgin would be possible with that. I am wondering about the cost of using sms though.

**Comment by Zontex** [comment]
> @Zontex rest/mqtt integration is very easily possible. Interacting with GPIO would also be easy. If you have questions ping me.
> 
> The new notification method is generic. So any plugin can trigger a notification and the delivery methods are defined per user. SMS, Telegram, or pidgin would be possible with that. I am wondering about the cost of using SMS though.

The cost for Twillio will be pretty unreasonable. what I do is I duplicate my SIM card (same account, different number) and it has an unlimited SMS package under my mobile provider package, this allows me to use it as much as I need.

I think no need for direct integration to GPIO because connecting the wires directly to raspberry pi (OmniKit in my case) makes it very messy, I tried. I power the entire system by PoE and keeping it next to the router makes it very convenient, controlling the LEDs through another microcontroller will be a better solution probably.

Regarding the integration part, sounds amazing, looks like the new plugin system fits perfectly for this kind of task :) Let me know if there is anything I can contribute to.

**Comment by matmair** [comment]
> The cost for Twillio will be pretty unreasonable. what I do is I duplicate my SIM card (same account, different number) and it has an unlimited SMS package under my mobile provider package, this allows me to use it as much as I need.

Ok, that makes more sense financially.

> Regarding the integration part, sounds amazing, looks like the new plugin system fits perfectly for this kind of task :) Let me know if there is anything I can contribute to.

If you have time - writing a plugin to test how good the docs and APIs would be tremendous!

@Zontex Also pulling and reviewing #2372 would be a great help.

**Reply by Zontex** [reply]
Where can I find the documentation for the new plugin system? would love to give it a go and update the results

**Reply by matmair** [reply]
@zontex There is a whole section in the docs:
https://inventree.readthedocs.io/en/latest/extend/how_to_plugin/
